### ISSUE TITLE
HashMap struct key with 'static lifetime screws with return lifetime using methods like `.get()`

### ISSUE THREAD

**Issue by jsdw** [rust-lang/rust#124614]
I'm not sure what words to use to actually describe (or search for a dupe) this issue, so I'll just post it here and hope that it makes more sense to you guys!

So basically, if I write this code:

```rust
use std::collections::HashMap;

#[derive(PartialEq,Eq,Hash)]
struct Key<'a> {
    key: &'a str
}

struct Foo {
    types: HashMap<Key<'static>, String>,
}

impl Foo {
    fn resolve<'this, 'a>(
        &'this self,
        key: &'a str,
    ) -> Option<&'this String> {
        self.types.get(&Key { key })
    }
}
```

I get back this error:

```rust
error: lifetime may not live long enough
  --> src/lib.rs:17:9
   |
13 |     fn resolve<'this, 'a>(
   |                -----  -- lifetime `'a` defined here
   |                |
   |                lifetime `'this` defined here
...
17 |         self.types.get(&Key { key })
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ method was supposed to return data with lifetime `'this` but it is returning data with lifetime `'a`
   |
   = help: consider adding the following bound: `'a: 'this`
```

Despite the fact that `HashMap::get()` returns the values with a lifetime of self, and nothing to do with the key provided. 

However, this _does_ work:

```rust
use std::collections::HashMap;

#[derive(PartialEq,Eq,Hash)]
struct Key<'a> {
    key: &'a str
}

struct Foo {
    types: HashMap<&'static str, String>,
}

impl Foo {
    fn resolve<'this, 'a>(
        &'this self,
        key: &'a str,
    ) -> Option<&'this String> {
        self.types.get(key)
    }
}
```

Removing the `struct` indirection when defining the key type seems to make things happy!

## Background

The above is a minimal reproduction of the issue I ran into, but the reason I ran into it is because I want to have something like:

```rust
use std::collections::HashMap;

struct Foo {
    types: HashMap<(String,String), String>,
}

impl Foo {
    fn resolve<'this, 'a>(
        &'this self,
        key1: &'a str,
        key2: &'a str,
    ) -> Option<&'this String> {
        self.types.get(&(key1, key2))
    }
}
```

In other words, I want to be able to borrow more than 1 value to lookup some owned key in the hashmap. I went down the route of trying `struct Key<'a> { a: Cow<'a,str>, b: Cow<'a,str>` and then `HashMap<Key<'static>, String>`, but this led me to bumping up against this odd lifetime issue :)

Any pointers would be amazingly welcome!

**Comment by pacak** [comment]
I simplified the example a bit, nethier `Foo` nor `Key` are needed.

```rust
use std::collections::HashMap;

fn resolve<'this, 'a, T>(types: &'this HashMap<&'static str, T>, key: &'a str) -> Option<&'this T> {
    types.get(&key)
}
```

**Comment by jsdw** [comment]
Thank you for the very insightful response @HeroicKatora!

As I switched to using `hashbrown::HashMap` for no-std support, I ended up finding the `raw_entry` API really useful for working around the issue and did something like this, which I'll leave here just in case it helps anybody:

```rust
// Define a key type to insert into a hashbrown::HashMap
#[derive(Clone, Debug, PartialEq, Eq)]
struct Key {
    a: String,
    b: String,
}

impl core::hash::Hash for Key {
    fn hash<H: core::hash::Hasher>(&self, state: &mut H) {
        hash_key(self.a, &self.b, state);
    }
}

fn hash_key<H: core::hash::Hasher>(a: &str, b: &str, state: &mut H) {
    a.hash(state);
    b.hash(state);
}

// And then, we can lookup using the raw_entry API like:
fn lookup<'map>(
    a: &str,
    b: &str,
    map: &'map HashMap<Key, Info>,
) -> Option<(&'map Key, &'map Info)> {
    // Manually construct the key hash
    let hash = {
        use core::hash::{BuildHasher, Hasher};
        let mut state = map.hasher().build_hasher();
        hash_key(a, b, &mut state);
        state.finish()
    };

    // Look up the entry by hash, and then compare keys we get back to find the matching key.
    map.raw_entry().from_hash(hash, |k| k.a == a && k.b == b)
}
```

Though at some point I might go back to using `struct Key { a: Cow<'static,str>, b: Cow<'static,str> }` plus your wrapper approach instead, which would avoid any hashing shenanigans (and the chance for something to go wrong there if fields are added or whatever).

Happy to see this closed anyways now. Thanks!
